pkg/storegateway/series_chunks.go

```diff
@@ -335,36 +420,208 @@ func (c *loadingSeriesChunksSetIterator) Next() (retHasNext bool) {
+					groupChunks[k].Raw = &storepb.Chunk{}
+				}
+			}
 
-			err := c.chunkReaders.addLoad(chunk.blockID, chunk.ref, i, j)
+			ok, lastChkLen, err := parseGroup(loadedGroups[currentGroupIdx], groupChunks)
 			if err != nil {
-				c.err = errors.Wrap(err, "preloading chunks")
-				return false
+				return nil, fmt.Errorf("parsing chunk group (block %s, first ref %d, num chunks %d): %w", g.blockID, g.firstRef(), len(g.chunks), err)
+			}
+			if !ok {
+				// We estimate the length of the last chunk of a series.
+				// Unfortunately, we got it wrong. We need to refetch the whole group.
+				// We set the length correctly because we now know it.
+				unloadedS.groups[j].chunks[len(unloadedS.groups[j].chunks)-1].length = lastChkLen
+				underfetchedGroups = append(underfetchedGroups, underfetchedGroupIdx{
+					seriesIdx:      i,
+					groupIdx:       j,
+					firstChkIdx:    populatedChksCount,
+					loadedGroupIdx: currentGroupIdx,
+					blockID:        g.blockID,
+				})
 			}
+			populatedChksCount += len(g.chunks)
+			currentGroupIdx++
 		}
 	}
+	return underfetchedGroups, nil
+}
 
-	// Create a batched memory pool that can be released all at once.
-	chunksPool := pool.NewSafeSlabPool[byte](chunkBytesSlicePool, chunkBytesSlabSize)
+// parseGroup parses the byte slice as concatenated encoded chunks. lastChunkLen is non-zero when allChunksComplete==false.
+// An error is returned when gBytes are malformed or when not only the last chunk is incomplete.
+func parseGroup(gBytes []byte, chunks []storepb.AggrChunk) (allChunksComplete bool, lastChunkLen int64, _ error) {
+	for i := range chunks {
+		chunkDataLen, n := binary.Uvarint(gBytes)
+		if n == 0 {
+			return false, 0, fmt.Errorf("not enough bytes to infer length of chunk %d/%d", i, len(chunks))
+		}
+		if n < 0 {
+			return false, 0, fmt.Errorf("chunk length doesn't fit into uint64 %d/%d", i, len(chunks))
+		}
+		// ┌───────────────┬───────────────────┬──────────────┬────────────────┐
+		// │ len <uvarint> │ encoding <1 byte> │ data <bytes> │ CRC32 <4 byte> │
+		// └───────────────┴───────────────────┴──────────────┴────────────────┘
+		totalChunkLen := n + 1 + int(chunkDataLen) + crc32.Size
+		if totalChunkLen > len(gBytes) {
+			if i != len(chunks)-1 {
+				return false, 0, fmt.Errorf("underfetched before the last chunk, don't know what to do (chunk idx %d/%d, fetched %d/%d bytes)", i, len(chunks), len(gBytes), totalChunkLen)
+			}
+			return false, int64(totalChunkLen), nil
+		}
+		c := rawChunk(gBytes[n : n+1+int(chunkDataLen)])
+		if cEnc := c.Encoding(); cEnc != chunkenc.EncXOR {
+			return false, 0, fmt.Errorf("encoding (%d, %s) isn't XOR, don't know what to do ", cEnc, cEnc.String())
+		}
+		chunks[i].Raw.Type = storepb.Chunk_XOR
+		chunks[i].Raw.Data = c.Bytes()
+		// We ignore the crc32 because we assume that the chunk didn't get corrupted.
+		// TODO maybe check the crc for every 1 in 100 chunks? 1 in 1000?
```

Comment:
Thoughts? I'm not sure what guarantees us that the bytes we get from memcached and the bucket will not be corrupted. Probably TCP?